### DISCUSSION TITLE
why SpanGroup?

### DISCUSSION THREAD

**Discussion by afriedman412** [explosion/spaCy#12675] [Help: Best practices]
Is there any reason to use a `SpanGroup` instead of just a list of `Spans` if it isn't explicitly required by the code?

**Comment by jkgenser** [comment]
The purpose is that you may need to disambiguate separate groups. It's a way of keeping groups of spans organized. E.g. there may be one spacy component that creates spangroups under key `foo` and another component that creates spangroups under key `bar`. SpanGroup allows you to distinguish.

**Comment by adrianeboyd** [comment]
The main motivation for `SpanGroup` was to have a data structure containing spans that can be saved efficiently as part of the `Doc`, with `Doc.to_bytes` or `DocBin` or `pickle`.

It's extremely inefficient to save independent `Span` objects, since they'd each need a full doc+vocab to have all the information they'd need. spacy actually prevents you from pickling spans because it's not a good idea, with this error:

```
File "spacy/tokens/span.pyx", line 206, in spacy.tokens.span.Span.__reduce__
NotImplementedError: [E112] Pickling a span is not supported, because spans are only views of the parent Doc and can't exist on their own. A pickled span would always have to include its Doc and Vocab, which has practically no advantage over pickling the parent Doc directly. So instead of pickling the span, pickle the Doc it belongs to or use Span.as_doc to convert the span to a standalone Doc object.
```

In a script where you're just temporarily working with spans and you don't need to save them, then a list of spans is fine.

**Reply by afriedman412** [reply]
awesome thanks!!!
